### PR TITLE
add 0x gasless swaps

### PR DESCRIPTION
**Pull request by mintdart**

### CODE REVIEW

**Review by 0xngmi — CHANGES_REQUESTED** [review]
will this work for tokens like USDT that dont have EIP712 support? doesnt look like it, but matcha supports it

**Comment by mintdart** [comment]
> will this work for tokens like USDT that don't have EIP712 support? doesn't look like it, but matcha supports it

I tested a swap from USDT to DAI on polygon, their api doesn't return any signature errors

Right now my test swap fails because of `order not fillable` , will try to swap on mainnet
Screenshot

File: src/components/Aggregator/adapters/0xGasless.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+export const isGasless = true;
+
+export async function getQuote(chain: string, from: string, to: string, amount: string, extra) {
+	// amount should include decimals
+
+	const tokenFrom = from === ethers.constants.AddressZero ? nativeToken : from;
+	const tokenTo = to === ethers.constants.AddressZero ? nativeToken : to;
+	const amountParam =
+		extra.amountOut && extra.amountOut !== '0' ? `buyAmount=${extra.amountOut}` : `sellAmount=${amount}`;
```
Comment (0xngmi) [inline src/components/Aggregator/adapters/0xGasless.tsx:26]: if its possible to trade with output amount then you must set `export const isOutputAvailable = true;`

**Comment by mintdart** [comment]
> main thing missing is to separate the approval process, so its in two steps (rather than 2 popups that pop automatically) and user can choose whether to approve infinite or approve only amount needed in permits

updated

File: src/components/Aggregator/adapters/0xGasless.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+				'0x-chain-id': chainToId[chain]
+			}
+		}
+	).then((r) => r.json());
+
+	// do not show quote if there's not enough liquidity
+	if (!data.liquidityAvailable) return null;
+
+	return {
+		amountReturned: data?.buyAmount || 0,
```
[the remark this answers is not in the document]
Comment (mintdart) [reply src/components/Aggregator/adapters/0xGasless.tsx:46]: copied it from 0x adapter, removed it

File: src/components/Aggregator/adapters/0xGasless.tsx (hunk 1 above)
Comment (mintdart) [reply src/components/Aggregator/adapters/0xGasless.tsx:26]: updated

**Review by 0xngmi — DISMISSED** [review]
minor comments, besides that leaving here two reminders for when it goes live:
- need to set it as one of those adapters where we query our backend always
- move api key to env var & replace it

also, is it possible to cancel 0x gasless orders like how you can cancel cowswap ones?

File: src/components/Aggregator/adapters/0xGasless.tsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	return body;
+}
+export async function swap({ signTypedDataAsync, rawQuote, chain, approvalData }) {
+	// user must sign approval.eip712 and the trade.eip712 objects returned by /quote for gasless approvals
+	const body = { ...(approvalData ?? {}) };
+
+	const tradeSignature = await signTypedDataAsync({
+		domain: rawQuote.trade.eip712.domain,
```
[the remark this answers is not in the document]
Comment (mintdart) [reply src/components/Aggregator/adapters/0xGasless.tsx:141]: added https://github.com/LlamaSwap/interface/pull/266/commits/2aaf0b9afed530bd78a0610ae0ffe74a9224a9b6

File: src/components/Aggregator/index.tsx [hunk 4]
```diff
@@ -686,8 +696,16 @@ export function AggregatorContainer({ tokenList, sandwichList }) {
 					.div(100)
 					.toFixed(0)
 			: selectedRoute?.fromAmount;
+
+	const isGaslessApproval =
+		selectedRoute?.isGasless &&
+		(selectedRoute.price.rawQuote as any).approval.isRequired &&
```
[the remark this answers is not in the document]
Comment (mintdart) [reply src/components/Aggregator/index.tsx]: updated in https://github.com/LlamaSwap/interface/pull/266/commits/d772194a8d7d82db00bc41e522d6d214592e6cfc

**Comment by mintdart** [comment]
> minor comments, besides that leaving here two reminders for when it goes live:
> 
> * need to set it as one of those adapters where we query our backend always
> * move api key to env var & replace it
> 
> also, is it possible to cancel 0x gasless orders like how you can cancel cowswap ones?

updated code to query our backend always and use api key in env var

currently it's not possible to cancel orders by external party

**Review by 0xngmi — COMMENTED** [review]
hmm no retries on pending txs? or did i miss it?

**Comment by mintdart** [comment]
> hmm no retries on pending txs? or did i miss it?

added in https://github.com/LlamaSwap/interface/pull/266/commits/993f58be9f545773be21bd41b4feb864987f63ce

File: src/components/Aggregator/index.tsx [hunk 5]
```diff
@@ -1130,13 +1214,22 @@ export function AggregatorContainer({ tokenList, sandwichList }) {
 												/>
 											) : (
 												<Button
-													isLoading={swapMutation.isLoading || isApproveLoading}
-													loadingText={isConfirmingApproval ? 'Confirming' : 'Preparing transaction'}
+													isLoading={swapMutation.isLoading || isApproveLoading || gaslessApprovalMutation.isLoading}
+													loadingText={
+														isConfirmingApproval || selectedRoute?.isGasless || gaslessApprovalMutation.isLoading
```
Comment (0xngmi) [inline src/components/Aggregator/index.tsx:1219]: won't `selectedRoute?.isGasless` always be true when on a gasless route?

File: src/components/Aggregator/index.tsx [hunk 6]
```diff
@@ -1347,10 +1449,21 @@ export function AggregatorContainer({ tokenList, sandwichList }) {
 															/>
 														) : (
 															<Button
-																isLoading={swapMutation.isLoading || isApproveLoading}
-																loadingText={isConfirmingApproval ? 'Confirming' : 'Preparing transaction'}
+																isLoading={
+																	swapMutation.isLoading || isApproveLoading || gaslessApprovalMutation.isLoading
+																}
+																loadingText={
+																	isConfirmingApproval || selectedRoute?.isGasless || gaslessApprovalMutation.isLoading
```
Comment (0xngmi) [inline src/components/Aggregator/index.tsx:1456]: same here

File: server/src/submitSwap.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import { adapters } from './dexAggregators/list';
+
+const handler = async (event: AWSLambda.APIGatewayEvent): Promise<any> => {
+	const { protocol, chain } = event.queryStringParameters!;
+	const body = JSON.parse(event.body!);
+	const agg = adapters.find((ag) => decodeURIComponent(ag.name) === protocol);
```
[the remark this answers is not in the document]
Comment (0xngmi) [reply server/src/submitSwap.ts:6]: turns out that AWS API Gateway always decodes querystring params, so this is not needed at all

[the remark this answers is not in the document]
Comment (0xngmi) [reply server/src/submitSwap.ts]: removed decoding

File: server/src/submitSwap.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+			headers: {
+				'Cache-Control': `max-age=${3600}`,
+				'Access-Control-Allow-Origin': '*'
+			}
+		};
+	}
+	const res = await (agg as any).submitSwap({ chain, body });
+	return {
+		statusCode: 200,
+		body: JSON.stringify(res)
```
[the remark this answers is not in the document]
Comment (0xngmi) [reply server/src/submitSwap.ts]: added headers

File: src/components/Aggregator/index.tsx (hunk 5 above)
Comment (mintdart) [reply src/components/Aggregator/index.tsx]: correct, removed in https://github.com/LlamaSwap/interface/pull/266/commits/6b27c11a410773f517c9c54c9e4b4e51e16c49b1

**Review by 0xngmi — DISMISSED** [review]

**Pull request merged**